Header visualization

vincerubinetti opened the pull request:
- disable logo animation replaying (because it interferes with the new viz)
- make new background network visualization for header using canvas (for speed), and incorporating a nice animation and 3d rotation with mouse/touch. scales to fit pixel density of screen.
- add mocks for canvas and other window/browser functions for tests to work

falquaddoomi commented on src/util/math.ts:
```diff
@@ -0,0 +1,13 @@
+// euclidean distance between two points
+export const dist = (x1 = 0, y1 = 0, x2 = 0, y2 = 0): number =>
```
Just out of curiosity, why specify default values for this file's function's arguments?

vincerubinetti replied:
It's an easy/terse way to have typescript infer that they are number types, and doubles as a safe guard. At one point in my experimentation/iteration I was also using it like `dist(x,y)` to get the distance of a point to the origin.

falquaddoomi reviewed (approved):
This looks super! I'm impressed. :)

Also, I didn't see anything of concern; apparently the canvas functions you're using are broadly compatible (according to caniuse.com). Nice work!

kshefchek reviewed (approved):
this is awesome!

vincerubinetti commented:
@falquaddoomi I've once again made changes after review. Please take a look at the last two commits.

In the first commit, I've removed the dot random motion ("spinning") because I realized with the 3d effect it was unnecessary and overwhelming.

In the second commit, I removed the `tinycolor` library and implemented basic color implementation myself. I noticed that the `step()` loop was taking about `8ms` on my machine. Note that 60fps leaves `16ms` in between frames, and the step call should take up as little as possible to make sure the rest of the page is responsive. Of that `8ms`, < `1ms` was for `draw()`, < `1ms` was for the math/projection/3d portion of `move()` and > `5ms` was taken by the simple color blending portion of `move()`. This was probably due to parsing hex string, blending, then re-stringifying. Instead, now colors are just stored as a basic rgb tuple, and now the color portion also takes < `1ms`, and the whole `step()` loop takes ~`1ms`.  Conservatively assuming older hardware runs 10x slower than my average Surface Laptop from 2 years ago, that's `10ms` and still leaves enough room within the `16ms`.

falquaddoomi commented on src/components/TheViz.vue:
```diff
@@ -0,0 +1,257 @@
+        // only link if dots close enough together
+        if (
+          dist(from.point.x, from.point.y, to.point.x, to.point.y) <
+          gap * 1.5
```
Since, you're interested in optimization, you might consider adding a `distSq()` function that doesn't call `Math.sqrt()`, which is typically slow. You'd replace the above with something like `distSq(...) < Math.pow(gap * 1.5, 2)`.

I don't know how much of a difference it makes nowadays, but it was a popular trick back in the day.

vincerubinetti replied:
Cool trick, I've never heard of that one. I'll keep it in mind for the future if we find it running unacceptably slow on an old device, but I imagine v8 and other js engines are probably so optimized today that it's not needed? Also sub 1ms is good with me, and dist is only a small part of everything being done. Note that it's only being called in `generate()` (which only runs once on page load or resize, debounced) and `pulse()` (which runs every once 5 seconds). So given that I'd like to keep it the recognizable distance formula.

\* "once" one frame, once per dot.
